### PR TITLE
fix: sanity check events of ZetaSent/ZetaReceived/ZetaReverted/Withdrawn/Deposited

### PR DESCRIPTION
**Pull request by ws4charlie**
# Description

1. Added sanity check function for events of `ZetaSent`/`ZetaReceived`/`ZetaReverted`/`Withdrawn`/`Deposited`.
2. Applied the checks in both inTx/outTx observation and inTx tracker handling.

Closes: [1537](https://github.com/zeta-chain/node/issues/1537)
Closes: [418](https://github.com/code-423n4/2023-11-zetachain-findings/issues/418)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [x] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: zetaclient/evm_client.go [hunk 1]
```diff
@@ -57,7 +57,12 @@ type EVMLog struct {
 }
 
 const (
-	DonationMessage = "I am rich!"
+	DonationMessage    = "I am rich!"
+	TopicsZetaSent     = 3 // https://github.com/zeta-chain/protocol-contracts/blob/d65814debf17648a6c67d757ba03646415842790/contracts/evm/ZetaConnector.base.sol#L34
+	TopicsZetaReceived = 4 // https://github.com/zeta-chain/protocol-contracts/blob/d65814debf17648a6c67d757ba03646415842790/contracts/evm/ZetaConnector.base.sol#L45
+	TopicsZetaReverted = 3 // https://github.com/zeta-chain/protocol-contracts/blob/d65814debf17648a6c67d757ba03646415842790/contracts/evm/ZetaConnector.base.sol#L54
+	TopicsWithdrawn    = 3 // https://github.com/zeta-chain/protocol-contracts/blob/d65814debf17648a6c67d757ba03646415842790/contracts/evm/ERC20Custody.sol#L43
+	TopicsDeposited    = 2 // https://github.com/zeta-chain/protocol-contracts/blob/d65814debf17648a6c67d757ba03646415842790/contracts/evm/ERC20Custody.sol#L42C1-L42C1
```
Comment (lumtis) [inline zetaclient/evm_client.go:65]: Can we add a comment explaining what these represent?

File: zetaclient/utils.go [hunk 2]
```diff
@@ -170,6 +170,32 @@ func (t *DynamicTicker) Stop() {
 	t.impl.Stop()
 }
 
+// CheckEvmTxLog checks the basics of an EVM tx log
+func (ob *EVMChainClient) CheckEvmTxLog(vLog *ethtypes.Log, wantAddress ethcommon.Address, wantHash string, wantTopics int) error {
```
Comment (lumtis) [inline zetaclient/utils.go:174]: Can we add tests for this method?

File: zetaclient/evm_client.go [hunk 3]
```diff
@@ -977,12 +998,14 @@ func (ob *EVMChainClient) observeERC20Deposited(startBlock, toBlock uint64) uint
...
-		if !iter.Event.Raw.Removed && iter.Event.Raw.BlockNumber > 0 { // skip if chain reorg removed this event
+		// sanity check tx event
+		err := ob.CheckEvmTxLog(&iter.Event.Raw, addrCustody, "", TopicsDeposited)
+		if err == nil {
 			events = append(events, iter.Event)
 			continue
 		}
-		ob.logger.ExternalChainWatcher.Warn().Msgf("observeERC20Deposited: invalid event in tx %s at height %d for chain %d",
-			iter.Event.Raw.TxHash.Hex(), iter.Event.Raw.BlockNumber, ob.chain.ChainId)
+		ob.logger.ExternalChainWatcher.Error().Err(err).Msgf("observeERC20Deposited: invalid Deposited event in tx %s on chain %d at height %d",
```
Comment (lumtis) [inline zetaclient/evm_client.go:1007]: I would maybe keep a `Warn` log for these kind of error since it might be expected

Comment (ws4charlie) [reply zetaclient/evm_client.go:1007]: okay. I can set it back to `Warn` as DataDog still prints it out and we can use it for investigation.

File: zetaclient/evm_client.go (hunk 1 above)
Comment (ws4charlie) [reply zetaclient/evm_client.go:65]: sure

Comment (ws4charlie) [reply zetaclient/evm_client.go]: [commit](https://github.com/zeta-chain/node/pull/1539/commits/cb257f672246dee79554c565f711dec5b3ab1345)

File: zetaclient/utils.go (hunk 2 above)
Comment (ws4charlie) [reply zetaclient/utils.go:174]: same [commit](https://github.com/zeta-chain/node/pull/1539/commits/cb257f672246dee79554c565f711dec5b3ab1345)

File: zetaclient/evm_client.go (hunk 3 above)
Comment (ws4charlie) [reply zetaclient/evm_client.go]: same [commit](https://github.com/zeta-chain/node/pull/1539/commits/cb257f672246dee79554c565f711dec5b3ab1345)

**Review by lumtis — APPROVED** [review]

**Review by kingpinXD — APPROVED** [review]

**Review by brewmaster012 — APPROVED** [review]
LGTM

**Pull request merged**